crates/sui-source-validation/src/lib.rs

```diff
@@ -396,3 +474,273 @@ fn local_modules(
+                    "REQUIRE".bold().green(),
+                    toolchain_version.compiler_version.yellow(),
+                );
+                package_version_map.insert(*package, toolchain_version);
+            }
+        }
+    }
+
+    let mut units = vec![];
+    let mut package_compiled: HashSet<Symbol> = HashSet::new();
+    // Iterate over compiled units, and determine whether they need to be recompiled and replaced by a prior compiler.
+    for (package, local_unit) in compiled_units {
+        match package_version_map.get(package) {
+            Some(ToolchainVersion {
+                compiler_version, ..
+            }) if compiler_version == CURRENT_COMPILER_VERSION => {
+                units.push((*package, local_unit.clone()));
+                continue;
+            }
+            Some(toolchain_version) if !package_compiled.contains(package) => {
+                let package_root = SourcePackageLayout::try_find_root(&local_unit.source_path)?;
+                download_and_compile(package_root.clone(), toolchain_version, package)?;
+                package_compiled.insert(*package);
+
+                let compiled_unit_paths = vec![package_root.clone()];
+                let compiled_units = find_filenames(&compiled_unit_paths, |path| {
+                    extension_equals(path, MOVE_COMPILED_EXTENSION)
+                })?;
+                let build_path = package_root.clone().join("build").join(package.as_str());
+                // If we compiled units with a previous compiler, add *all* those units at once, so that
+                // we don't need to resolve them by path one-by-one.
+                for bytecode_path in compiled_units {
+                    info!("bytecode path {bytecode_path}, {package}");
+                    let local_unit =
+                        decode_bytecode_file(build_path.clone(), package, &bytecode_path)?;
+                    units.push((*package, local_unit))
+                }
+                continue;
+            }
+            Some(_) => {
+                // If a toolchain version for a previous compiler exists, we
+                // already added all compiled units to deps_compiled_units.
+                continue;
+            }
+            None => bail!("Expected a version for {package} but none found"),
+        }
+    }
+    Ok(units)
+}
+
+fn download_and_compile(
+    root: PathBuf,
+    ToolchainVersion {
+        compiler_version,
+        edition,
+        flavor,
+    }: &ToolchainVersion,
+    dep_name: &Symbol,
+) -> anyhow::Result<()> {
+    let binaries_path = &*MOVE_HOME; // E.g., ~/.move/binaries
+    let mut dest_dir = PathBuf::from(binaries_path);
+    dest_dir = dest_dir.join("binaries");
+    let dest_version = dest_dir.join(compiler_version.clone());
+    let platform = detect_platform()?;
+    let dest_binary = dest_version.join(format!("target/release/sui-{}", platform));
+    let dest_binary_os = OsStr::new(dest_binary.as_path());
```

Comment:
nit: You might be able to clean-up the path creation logic a bit using `extend`/`from_iter`/`push` and friends:
```suggestion
    let dest_dir = PathBuf::from_iter([&*MOVE_HOME, "binaries"]);
    let dest_version = dest_dir.join(&compiler_version);
    let platform = detect_platform()?;
    let mut dest_binary = dest_version.clone()
    dest_binary.extend(["target", "release", &format!("sui-{platform}")]);
    let dest_binary_os = OsStr::new(dest_binary.as_path());
```